### PR TITLE
file middleware

### PR DESCRIPTION
**Pull request by tanner0101**
- [x] fixes viewrenderer `.make()`
- [x] fixes some bugs in http serialization
- [x] adds streaming file reader
- [x] adds file exists checker
- [x] adds file middleware
- [x] adds req.streamFile helper

### CODE REVIEW

File: Sources/Core/File/File.swift [hunk 1]
```diff
@@ -20,8 +21,13 @@ public final class File: FileReader, FileCache {
     }
 
     /// See FileReader.read
-    public func read(at path: String) -> Future<Data> {
-        let promise = Promise(Data.self)
+    public func read<S>(at path: String, into stream: S, chunkSize: Int = 2048)
+        where S: Async.InputStream, S.Input == Data
```
Comment (Joannis) [inline Sources/Core/File/File.swift]: We really should not read files in chunks of `Data`. They should pass `ByteBuffer`s for performance (less copies) and uniformity with the rest of Vapor 3's APIs that work with binary

File: Sources/Core/File/FileCache.swift [hunk 2]
```diff
@@ -5,24 +5,22 @@ import Foundation
 public protocol FileCache {
     /// Fetches the file from the cache
     /// Supply a queue to complete the future on.
-    func getFile<H: Hashable>(hash: H) -> Future<Data?>
+    func getCachedFile(at path: String) -> Data?
```
Comment (Joannis) [inline Sources/Core/File/FileCache.swift]: A note in general.

I think that if we do this right, we (almost) never want to cache files in-memory since it needs to be cached per-thread (so 1MB of cached data becomes more than just 1MB of memory usage). So we should really try to use this as little as possible.

File: Sources/HTTP/Serializer/ChunkEncoder.swift [hunk 3]
```diff
@@ -9,39 +9,40 @@ final class ChunkEncoder: Async.Stream, ClosableStream {
...
-    let stream = BasicStream<Output>()
+    let outputStream = BasicStream<Output>()
     
     /// See OutputStream.onOutput
     func close() {
         // An empty padding chunk
-        self.onInput(ByteBuffer(start: nil, count: 0))
-        
-        stream.close()
+        let header = Array(String.init(0, radix: 16, uppercase: true).utf8) + crlf
```
Comment (Joannis) [inline Sources/HTTP/Serializer/ChunkEncoder.swift]: This can be a `fileprivate` global variable for this array + clrfs.

File: Sources/Vapor/Engine/Request+StreamFile.swift [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    /// to the public directory, ie: `drop.publicDir + "myFile.cool"`
+    ///
+    /// If none match represents an ETag that will be used to check if the file has
+    /// changed since the last load by the client. This allows clients like browsers
+    /// to cache their files and avoid downloading resources unnecessarily.
+    /// Most often calculated w/
+    /// https://tools.ietf.org/html/rfc7232#section-3.2
+    ///
+    /// For an example of how this is used, look at 'FileMiddleware'
+    public func streamFile(at path: String) throws -> Response {
```
Comment (Joannis) [inline Sources/Vapor/Engine/Request+StreamFile.swift]: This should be a static function on `Response`. This function is unrelated to the current request

**Comment by kdawgwilk** [comment]
I attempted to use this branch and ran into a few issues:
1) FileMiddleware runs before routing so tries to load my route as a file which errors out
  - This was done in Vapor 2 by using a try catch in the middleware to watch for a route not found error and only trying to serve up files when they weren't handled by the router
2) Content-Type header for all files is `text/html` which causes css/js files to not load in browser apparently

**Comment by tanner0101** [comment]
@kdawgwilk what error specifically was getting thrown? i'm not seeing that problem. the middleware checks if the file exists first before attempting to load it so that shouldn't be happening.

File: Sources/Core/File/FileCache.swift (hunk 2 above)
Comment (tanner0101) [reply Sources/Core/File/FileCache.swift]: this is only used for leaf templates so far, where the small file size of the templates shouldn't be a big deal (not like caching images). but we should take another look at how we can optimize leaf in the future.

File: Sources/Vapor/Engine/Request+StreamFile.swift (hunk 4 above)
Comment (tanner0101) [reply Sources/Vapor/Engine/Request+StreamFile.swift]: it's not since we need the current request's event loop

File: Sources/Core/File/File.swift (hunk 1 above)
Comment (tanner0101) [reply Sources/Core/File/File.swift]: we should at least expose an _option_ for `Data` for our high levels APIs. it has much better usability. i'm fine w/ ByteBuffer being included though. just not sure which one should be the base.
